### PR TITLE
[CocoaPods] Make podspec great again.

### LINKED ISSUE #11272: Integration With Existing iOS Project failed: 'CSSLayout/CSSLayout.h' file not found
### Description

After upgrade to react-native@0.39, Integration With Existing Empty iOS Project failed, and old workaround `"postinstall": "find ./node_modules/react-native \\( -name *.h -o -name *.m \\) -print0 | xargs -0 sed -i '' -e 's:<CSSLayout/\\(.*\\)>:\"\\1\":g'"` doesn't work now

react-native@0.38 works just fine
### Reproduction

1. create an empty iOS project with swift
2. follow the official guide of integration with existing app
3. try run iOS app and then failed

### Additional Information

* React Native version: 0.39
* Platform: iOS
* Operating System: MacOS Sierra 10.12.1
* CocoaPods: 1.1.1

### LINKED ISSUE #11572: [0.40] Warning: Swift Projects Fail
With the new changes in e1577df and 59407f3, header references have been changed from quotes to angle brackets. While this may work in an Obj-C project, it causes several compilation errors for swift projects.

Previously to `0.40`, you could create a new swift project and perform the following steps to compile successfully:
* add the `-all_load` and `-lc++` linker flags
* add react native dir (recursively) to `Header Search Paths`
* setup your bridging header

Full steps for swift setup details can be seen [here](https://gist.github.com/boopathi/27d21956fefcb5b168fe).

Now according to the summary in update 59407f3:
 > $(BUILT_PRODUCTS_DIR)/include, which is added to the include path by Xcode by default

The above statement implies that customizing `Header Search Paths` is no longer necessary, but is this true for Swift? In fact it seems there's no longer *any* straightforward way for Swift projects to find these angle bracket headers when libraries are included directly in the project (not using cocoapods), similar to what has been detailed [on stack overflow](http://stackoverflow.com/questions/17465902/use-of-external-c-headers-in-objective-c). In addition to `Header Search Paths`, I've tried `User Header Search Paths` and setting `Always Search User Paths` to `Yes`. No combination thus far yields successful compilation, instead throwing these errors:

image

Why is this change being made? I've been using swift with RN successfully since the very beginning, and now there is risk of not being able to upgrade without jumping through hoops (i.e. npm post install scripts to rewrite headers). Please don't introduce such a breaking change unless there's a verified way of supporting swift!

### LINKED ISSUE #11721: Reliable build on `^0.39.2` fails when upgrading to `^0.40.0`
```bash
** BUILD FAILED **


The following build commands failed:
	CompileC /XXX-FULLPATH-XXX/ios/build/Build/Intermediates/RCTSettings.build/Debug-iphonesimulator/RCTSettings.build/Objects-normal/x86_64/RCTSettingsManager.o RCTSettingsManager.m normal x86_64 objective-c com.apple.compilers.llvm.clang.1_0.compiler
(1 failure)
```

### LINKED ISSUE #11781: RCTConvert.h yoga/Yoga.h file not found
I am logging this as a bug as it seems like a bunch of people have logged it as an issue on Stackoverflow ... but seems to be a bug http://stackoverflow.com/questions/41505616/yoga-yoga-h-header-file-not-found-after-react-native-integration-in-the-existing

Upgraded to latest react-native 0.4.0

Build project for iOS on Xcode and you get the error message: `/Users/adamginsburg/Documents/development/buzzy9/node_modules/react-native/React/Base/RCTConvert.h:13:9: 'yoga/Yoga.h' file not found`

* React Native version: 0.4.0
* Platform: iOS
* Operating System: The build issue happen on XCode 8

### PR DESCRIPTION
**Pull request by alloy**
Fixes #11272
Fixes #11572
Fixes #11781

The main changes here are:

* This depends on the latest CocoaPods (1.2.0). It’s currently in RC, but if I’m not mistaken a proper release is expected soon. /cc @dantoml
* Adds required header search paths for the jschelpers and cxxreact subspecs.
* Makes the jschelpers and cxxreact headers private to building React Native, not visible to the user’s project.
* It uses the canonical upstream Yoga v1.0.0 podspec: https://github.com/facebook/yoga/blob/master/Yoga.podspec
* Consistent styling.

I have been able to get our app to build again using this https://github.com/artsy/emission/pull/437. The spec has some warnings, but otherwise fully passes lint.

@rh389 @sjmueller Could you please test with your projects?

### CODE REVIEW

**Comment by alloy** [comment]
Next week I will take some time to get both Yoga and React Native to perform linting on CI, so that I can hear of these regressions when they happen. (Do you have some way with your bot to notify specific people when specific parts of the CI setup fail?)

File: React.podspec [hunk 1]
```diff
@@ -1,140 +1,138 @@
...
 
-  s.subspec 'Core' do |ss|
-    ss.dependency      'React/yoga'
-    ss.dependency      'React/cxxreact'
-    ss.source_files  = "React/**/*.{c,h,m,mm,S}"
-    ss.exclude_files = "**/__tests__/*", "IntegrationTests/*", "React/**/RCTTVView.*", "ReactCommon/yoga/*"
-    ss.frameworks    = "JavaScriptCore"
-    ss.libraries     = "stdc++"
+  s.subspec "Core" do |ss|
+    ss.dependency             "Yoga", "~> 1.0"
```
Comment (ide) [inline React.podspec]: Can you make this use the local copy of Yoga? The RN tests in this repo (and separate tests within FB I imagine) are run against whatever copy of Yoga is synced into this repo and we want to avoid divergence between using CocoaPods vs. manually linking the framework.

**Comment by rh389** [comment]
LGTM - working nicely in a mixed Swift/Obj-C project importing React as a framework, with a bunch of React subspecs and third-party native modules installed as pods. No hacks, no post-install scripts, no manual config! 🎉 

Thanks for finding the time to work on this 👍

File: React.podspec (hunk 1 above)
Comment (alloy) [inline React.podspec]: @ide Yeah you’re totally right that the code should match what’s being used in the tests, but I was under the impression that Yoga moving to an official podspec was paving the way for us to make using both RN and Yoga as dependencies of our apps nicer.

I.e. I would really like to have the choice to have flattened transitive dependencies in my app, no duplications (smaller binary size), and less dealing with possible duplicate symbols and what not.

@emilsjolander do you know if that was/is not the case?

----

Nonetheless, as it stands, the code is indeed out of sync, I should have noted that.

1. The code vendored in RN is [this Yoga version](https://github.com/facebook/yoga/commit/a85bd4ad2a25acd29f07d5ba1a87b747d9fb7546).
2. The published v1.0.0 of Yoga is [this newer revision](https://github.com/facebook/yoga/commit/8d74e01f41f062d732bdc4de34dddf213b0fb173).

So if this were to be used, either the version in 1 needs to be published as e.g. Yoga v0.9, or the vendored code in RN should be updated to the version in 2.

**Comment by javache** [comment]
Thanks for your help @alloy! We should definitely get this on CI as well, @hramos can you help with that?

File: React.podspec (hunk 1 above)
Comment (javache) [inline React.podspec]: Getting releases in sync will be very hard. @dshahidehpour, could we have the podspec in the dirsynced folder so we can reference the included copy?

File: React.podspec [hunk 2]
```diff
@@ -1,140 +1,138 @@
...
+    ss.source_files         = "React/**/RCTTVView.{h, m}"
   end
 
-  s.subspec 'jschelpers' do |ss|
-    ss.source_files = 'ReactCommon/jschelpers/{JavaScriptCore,JSCWrapper}.{cpp,h}'
-    ss.header_dir   = 'jschelpers'
+  s.subspec "jschelpers" do |ss|
+    ss.source_files         = "ReactCommon/jschelpers/{JavaScriptCore,JSCWrapper}.{cpp,h}"
+    ss.private_header_files = "ReactCommon/jschelpers/{JavaScriptCore,JSCWrapper}.h"
+    ss.pod_target_xcconfig  = { "HEADER_SEARCH_PATHS" => "$(PODS_TARGET_SRCROOT)/ReactCommon" }
```
Comment (javache) [inline React.podspec]: This will work for now, but it doesn't match the xcode behaviour in terms of recursion. For now, we can try not include any subfolders in these directors.

File: React.podspec (hunk 1 above)
Comment (alloy) [inline React.podspec]: > Getting releases in sync will be very hard.

Oh, bummer, I thought that cutting Yoga releases wasn’t a big deal [“If you need a new release to be cut please file a task.”](https://github.com/facebook/yoga/releases/tag/v2017.01.23.00)

> could we have the podspec in the dirsynced folder so we can reference the included copy?

You mean that users would add that to their Podfile separately? Because I don’t see a way to reference it from RN’s podspec 🤔

File: React.podspec (hunk 2 above)
Comment (alloy) [inline React.podspec]: You mean you would add subfolders but _not_ include them with the same path, but rather all under the ‘namespace’?

File: React.podspec (hunk 1 above)
Comment (alloy) [inline React.podspec]: Took a quick look at using the vendored Yoga code, alas this is not going to work well right now due to https://github.com/CocoaPods/CocoaPods/issues/6440

Comment (ide) [inline React.podspec]: The way I would be comfortable with relying with Yoga from the CocoaPods repo is if Facebook were to use the same snapshot of Yoga (whether via CocoaPods or another mechanism e.g. Buck). I strongly prefer a systemic way that makes sure FB engineers are less likely to break open-source and vice versa, by making sure everyone is running the same copy of Yoga.

Comment (alloy) [inline React.podspec]: I think settling on the same systems is going to be a hard sell. If I understand it right, there’s the following types of consumers:

* FB engineers, I assume they use Buck
* OSS CP users
* OSS npm users

What’s most important is that they can all know, at release time, what the right version is, no? So some metadata about the Yoga commit/tag that’s being used is the main requirement. This could be a file next to the vendored code, or maybe something like a git subtree (assuming that git records metadata about these like with subomdules, but I’ve never used them).

The same applies to other parts of RN, btw, like the snapshot (image) testing module. The way it’s currently included makes it hard to integrate with apps that use FBSnapshotTestcase directly.

Comment (ide) [inline React.podspec]: The systems don't need to be the same, I want just the Yoga code to be the same -- I would be onboard with specifying a precise commit.

**Comment by DanToml** [comment]
I just released 1.2.0 proper. 👍. https://github.com/CocoaPods/CocoaPods/releases/tag/1.2.0

**Comment by mkonicek** [comment]
Hi guys, what's the conclusion here? Is this good to go, needs changes, or there is no consensus on what to do next?

File: React.podspec [hunk 3]
```diff
@@ -1,140 +1,138 @@
...
+  s.homepage                = "http://facebook.github.io/react-native/"
+  s.license                 = package["license"]
+  s.author                  = "Facebook"
+  s.source                  = { :git => "https://github.com/facebook/react-native.git", :tag => "v#{s.version}" }
+  s.default_subspec         = "Core"
+  s.requires_arc            = true
+  s.platform                = :ios, "8.0"
+  s.pod_target_xcconfig     = { "CLANG_CXX_LANGUAGE_STANDARD" => "c++14" }
+  s.preserve_paths          = "package.json", "LICENSE", "LICENSE-CustomComponents", "PATENTS"
+  s.cocoapods_version       = ">= 1.2.0.beta.1"
```
Comment (bachand) [inline React.podspec]: @alloy is there a feature of CocoaPods 1.2.0 that is required to support this new podspec? If not, do you think we could relax the version requirement? Would be nice if it wasn't necessary to update to consume this fix!

Thanks for doing this, solves the problem we've been having!! 👍

**Comment by rh389** [comment]
@mkonicek I think ideally we need someone at FB to modify the process by which Yoga etc are copied in from other repos so that CP can know which version to pull to match the local copy.

@alloy's idea of generating a metadata file that sits alongside any imported code specifying at least a github tag/SHA (of the 'canonical' open-source repo) sounds pretty useful to me.

**Comment by dshahidehpour** [comment]
@emilsjolander is in charge of the Yoga mirror at FB.

**Comment by alloy** [comment]
@mkonicek Yup, what @rh389 said. I'll send my thoughts on how to do this to @emilsjolander tomorrow and will report back here once we've reached consensus.

File: React.podspec (hunk 3 above)
Comment (alloy) [inline React.podspec]: This version of the spec uses `POD_TARGET_SRCROOT`, which is a CocoaPods 1.2.0 only feature. I'll have another think when finalizing it.

File: React.podspec (hunk 2 above)
Comment (javache) [inline React.podspec]: In our xcode setup, a header like `ReactCommon/jschelpers/foo/bar.h` would be includable as `<jschelpers/bar.h>`, which won't work this setup as I understand it.

In the short term, we can avoid any adding new subfolders under these ReactCommon targets

**Comment by rops** [comment]
@alloy just tried this patch but #11721 doesn't seem to be fixed for me. I'm using latest versions of 3rd party modules ( with new imports <React/..> ) and I'm including React from pods. Building the 3rd party libraries fails at "<React/RCTBridgeModule> file not found". Any idea what's wrong? Maybe has to do with header search paths?

**Comment by alloy** [comment]
@javache I’ll spend some more time to see I can get that to work.

**Comment by alloy** [comment]
@rops Hmm, might be a different issue after all. I’ll take a good look at that one once this issue/PR is resolved.

**Comment by rops** [comment]
@alloy I've done some digging and it seems that the third party libraries can't find <React> because they don't have the Pods header path in their Header Search Path. Can you think of some cocoapods hackery to fix that?

**Comment by alloy** [comment]
@rops Let’s take this to #11721

**Comment by guysegal** [comment]
If this PR will be merged are we expecting to see it only in 0.43.0 (March release)?

**Comment by keatongreve** [comment]
@alloy This might need to be updated for some changes in master/0.42.x. I've been using the work in this PR to host versions of React in a private podspec, and while I got a working version of 0.41.0 pushed, 0.42.0-rc.1 was not successful (I didn't test rc.0).

This fixed it for me: https://github.com/hudl/react-native/commit/97f6565d2e8dc6954d4c0d97e9f8f2e03ed68b2b

Other than that, this work has been super helpful for how we're using React Native, hopefully we see this merged soon.

**Comment by alloy** [comment]
> For now, can we just have our own spec for Yoga as a subspec and not make use of the Yoga spec file?

@javache The problem here is that when building as a framework a header is expected to be part of the frameworks’s namespace, other namespaces are expected to be outside of the framework (i.e. in a different framework). Thus, an import like `yoga/Yoga.h` _cannot_ exist inside the same framework. As this import exists in public headers exported by React (`RCTConvert.h` and `RCTShadowView.h`), this leads to an error when trying to build a Clang module for the framework (or otherwise would have lead to an error if the user tried to import one of these headers).

In fact, the only reason that `jschelpers` and `cxxreact` can be subspecs at all is simply because I made their headers private to the project.

In conclusion the only possible interim solution I see is to add a separate podspec for the vendored Yoga code and have users add that one specifically to their Podfile. I’ll add an example to this PR.

**Comment by ide** [comment]
> In conclusion the only possible interim solution I see is to add a separate podspec for the vendored Yoga code and have users add that one specifically to their Podfile. I’ll add an example to this PR.

Sounds good. People already have to add subspecs for basic things like Images and Networking when including RN via CocoaPods, so I find it reasonable to ask the same for Yoga. It's extra work but not much extra complexity, if that makes sense.

**Comment by alloy** [comment]
@ide Good point.

**Comment by alloy** [comment]
Ok, pushed a podspec for the vendored Yoga and verified the following works in my setup:

```ruby
react_path = '../node_modules/react-native'
yoga_path = File.join(react_path, 'ReactCommon/yoga')

pod 'React', :path => react_path, :subspecs => ['RCTWebSocket']
pod 'Yoga', :path => yoga_path
```

File: React.podspec (hunk 3 above)
Comment (alloy) [inline React.podspec]: Alas, this is definitely the better fix for the majority of people. However, you could probably replace the `$(POD_TARGET_SRCROOT)` parts with `$(SRCROOT)` and point it to where your RN source lives relative to the Xcode source root.

**Comment by alloy** [comment]
@keatongreve Thanks, added 👍

**Comment by ide** [comment]
@allow Let's give this a shot but first can you update the `cocoapods_version` to require 1.2.0 (final) instead of the beta? 

Also I've opened up an issue to automatically vendor the Yoga podspec in the future but think that can wait.

**Comment by rh389** [comment]
After adding `pod 'Yoga', :path => '../node_modules/react-native/ReactCommon/yoga'` to my `Podfile` this update dropped in to my project nicely, with a clean install and build.

I don't normally use `Yoga` directly from my project, but I just tried `import yoga` from my `AppDelegate.swift` and made a couple of random yoga function calls - looks like that builds nicely too.

👍

**Comment by alloy** [comment]
Ok, I’ve bumped the CP requirement to a non-beta version and have additionally added a CI job that lints the specs, however that build can only be ran on _merged_ code, so it’s a no-op on PRs.

(Additionally, the same script can be used by those of us that prefer to publish the podspecs to private repos to do so in an automated fashion.)

**Comment by alloy** [comment]
I’ve removed #11721 from the list of fixed issues, I don’t believe it’s related after all.

**Comment by alloy** [comment]
@mkonicek If/when this PR gets merged, is it possible for you to setup an email rule to forward Travis failures in the podspecs build to me? I don’t believe there’s a way to do this in the Travis configuration.

**Comment by ide** [comment]
@facebook-github-bot shipit

**Comment by alloy** [comment]
Wow. You know what I only just now realised? The title of this PR… I wrote it while I was in the US and might have been watching news in the background and made a Freudian slip.

It was supposed to say “Make podspec **work** again.”

**Comment by guysegal** [comment]
It seem like that was you intention in the PR name.. Anyway you actually made podspec GREAT again.
We've been following this PR and it solves us a lot of problems!
Thank you

**Comment by Snorlock** [comment]
Im trying to make sense of this flow, does the ship it means that its merged? 😅 
When it was shipped, is it in the master or what branch can we use now, instead of the fork? 🤓

**Comment by rh389** [comment]
@Snorlock it's now in master (https://github.com/facebook/react-native/commit/031cb2045de0ed430d539eb3d5c44baa2fe3bc8b). I guess it'll probably be cherry-picked into 0.42 by the end of the month, otherwise it'll be in 0.43 at the end of March.

**Comment by Snorlock** [comment]
@rh389 cool thanks. One question though, both the fork, and now current master gives me errors in a completly new project on the android version.

```
import com.facebook.react.ReactApplication;
import com.facebook.react.ReactNativeHost;
```

These to imports cannot be resolved. When I switch to 0.41.2 its all ok. But this change is not documented anywhere. I think its weird that his has anything to do with android, but if anyone have any information?

**Comment by alloy** [comment]
@Snorlock It is unlikely that has anything to do with these changes, so you should probably look into other changes made since 0.41.2.

**Comment by Snorlock** [comment]
Yeah I tried using 0.42.0-rc.1, and it still works. But master breaks, so im creating a new issue. https://github.com/facebook/react-native/issues/12286 . I just asked if maybe you guys had an android project that you have experienced the same with :)

**Comment by huangcheng1** [comment]
Do you tried using Swift?

**Comment by alloy** [comment]
Yes, @rh389 did (see earlier comments). I also tested this as a framework build, which is all that is required for Swift interoperability.

**Comment by yimingtang** [comment]
@Snorlock

Because the `android` directory is empty in development mode and ignored by git. Pre-compiled Android binary exists in NPM releases only.

**Comment by Snorlock** [comment]
@yimingtang not completly sure what you mean? And is there a way to get around it?

**Comment by ide** [comment]
This PR has no effect on Android -- please try to keep the discussion at hand focused.

**Comment by yimingtang** [comment]
@Snorlock 

Version 0.41.2 and 0.42.0-rc.1 work because they contain pre-compiled React Android binary. Every time the react-native dev team releases a new version, they puts a new pre-compiled React Android ( `.aar`, `.jar` files) into the package. These files are under `android` directory. However, the directory is ignore by git (see the `.gitignore` file). Because the binary files can be generated from source files, the repository shouldn't track them.

If you are going to use the master branch. You'd better:

1. Fork the repo.
2. Checkout a new branch.
3. Compile React Android yourself.
4. Point react-native to the branch that you created.

More details: 
https://github.com/facebook/react-native/blob/master/Releases.md
https://github.com/facebook/react-native/blob/master/ReactAndroid/DevExperience.md

This is what my colleague did:
https://github.com/iException/react-native/tree/0.42-baixing
